### PR TITLE
fix: dtype cast error

### PR DESCRIPTION
**Pull request by yewentao256**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation and Context
fix dtype cast error

## Checklist
**Before PR**:

- [x] I have read and followed the workflow indicated in the [Contributors.md](https://github.com/DeepLink-org/DIOPI/blob/main/Contributors.md) to create this PR.
- [x] Pre-commit or linting tools indicated in [Contributors.md](https://github.com/DeepLink-org/DIOPI/blob/main/Contributors.md) are used to fix the potential lint issues.
- [x] Bug fixes are covered by unit tests, the case that causes the bug should be added in the unit tests.
- [x] New functionalities are covered by complete unit tests. If not, please add more unit test to ensure the correctness.
- [x] The documentation has been modified accordingly, including docstring or example tutorials.

**After PR**:

- [x] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

File: impl/camb/common/dtype_cast.cpp [hunk 1]
```diff
@@ -14,27 +14,30 @@
...
-    constexpr uint64_t uint8Bool = _MAKE_KEY(diopi_dtype_uint8, diopi_dtype_bool);
-    constexpr uint64_t int16Bool = _MAKE_KEY(diopi_dtype_int16, diopi_dtype_bool);
-    constexpr uint64_t int64Bool = _MAKE_KEY(diopi_dtype_int64, diopi_dtype_bool);
-    constexpr uint64_t int8Bool = _MAKE_KEY(diopi_dtype_int8, diopi_dtype_bool);
-    constexpr uint64_t int8Int64 = _MAKE_KEY(diopi_dtype_int8, diopi_dtype_int64);
-    constexpr uint64_t int64Int8 = _MAKE_KEY(diopi_dtype_int64, diopi_dtype_int8);
-
-    return boolInt64 == castType || int16Int64 == castType || uint8Bool == castType || int16Bool == castType || int64Bool == castType || int8Bool == castType ||
-           int8Int64 == castType || int64Int8 == castType;
+    constexpr std::array<uint64_t, 8> validCasts = {MAKE_KEY(diopi_dtype_bool, diopi_dtype_int64),
```
Comment (yangbofun) [inline impl/camb/common/dtype_cast.cpp:20]: 这个修改没有必要

File: impl/camb/common/dtype_cast.cpp [hunk 2]
```diff
@@ -43,45 +46,45 @@ static diopiError_t dataTypeCastTwice(diopiContextHandle_t ctx, DiopiTensor& des
...
     return diopiSuccess;
 }
 
 diopiError_t dataTypeCast(diopiContextHandle_t ctx, DiopiTensor& dest, const DiopiTensor& src) {
-    // check size of dest and src
-    DIOPI_CHECK(src.shape() == dest.shape(), "the shapes of src and dest are not equal");
-
     if (dest.dtype() == src.dtype()) {
-        if (dest.data() != src.data()) {
-            clone(ctx, src, dest);
```
Comment (yangbofun) [inline impl/camb/common/dtype_cast.cpp:81]: 这个地方不能删除

File: impl/camb/common/dtype_cast.cpp (hunk 1 above)
Comment (yewentao256) [reply impl/camb/common/dtype_cast.cpp:20]: 这个是考虑后这么写的，现在的那个可读性有点差，每次还需要声明变量int64Int8等。

File: impl/camb/common/dtype_cast.cpp (hunk 2 above)
Comment (yewentao256) [reply impl/camb/common/dtype_cast.cpp:81]: 这个应该我没删除，感觉是哪里rebase了，我恢复一下

Comment (yewentao256) [reply impl/camb/common/dtype_cast.cpp]: 1

File: impl/camb/common/dtype_cast.cpp (hunk 1 above)
Comment (yangbofun) [reply impl/camb/common/dtype_cast.cpp:20]: 这样改的话，既不是C写法，也不是C++的写法。两边的优点都没用到。

Comment (yewentao256) [reply impl/camb/common/dtype_cast.cpp:20]: 已修改

**Review by yangbofun — APPROVED** [review]

**Pull request merged**